woocommerce/google-listings-and-ads
# Add plugin version, Google Merchant Center account ID, and Google Ads account ID to all frontend tracking events

### Changes proposed in this Pull Request:

Project thread: pcTzPl-20H-p2

This PR implements the JS part of the project. The target is to attach the following properties when sending all JS tracking events:

- Plugin Version
- Google Merchant Center account ID if connected
- Google Ads account ID if connected

To achieve the above target, this PR:

- On the PHP side, add base tracking properties to the inline script of outputting the global JS variable `glaData`.
- Add a general state to this extension's wp-data store for accessing the prefetched data (`glaData`) so that the account IDs in frontend cache can be synced after connecting/disconnecting an account.
- To reduce confusion with the `recordEvent` function imported from `@woocommerce/tracks`
   - Rename the **js/src/utils/recordEvent.js** file to **tracks.js**.
   - In the **tracks.js**, add new functions to wrap the `recordEvent` and `queueRecordEvent` to attach the base tracking properties.
   - Replace all `recordEvent` calls with `recordGlaEvent`.
   - Replace all `queueRecordEvent` calls with `queueRecordGlaEvent`.
- Update tracking README.

💡 The `wcadmin_page_view` event is sent from Woo core, it will be handled by another PR

### Screenshots:

#### 📷 Base tracking properties

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/85e329cf-9ae9-4253-9009-e075ee8d602c)

#### 📹 Attach Google Merchant Center account ID if connected

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/e4d6180f-de45-440a-b593-1b3702fd850e

#### 📹 Attach Google Ads account ID if connected

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/ccb4371e-62f8-4f4f-9913-f0105fe1d0cd

#### Queue a tracking event

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/a5eafe7b-a3d1-4efe-bbcf-a15811e8d94a

### Detailed test instructions:

💡 Apart from the `wcadmin_page_view` event, all tracking events should have the base properties.

1. Enable tracking debug tool
   - Open the Console tab of the browser DevTool.
   - Change the filter level to "All levels".
   - Run `localStorage.setItem( 'debug', 'wc-admin:*' )` and refresh the page to make it work.
   - It should be able to view tracking via the Console tab. Ref:
      ![1](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/698f85d8-e890-4ab6-939b-738acee6098f)
2. Disconnect Google Merchant Center and Google Ads accounts if already connected.
3. Go to the first step of onboarding flow.
4. Check if all tracking events have the `gla_version` property.
5. Check if all tracking events don't have the `gla_mc_id` property before connecting a GMC account.
6. Check if all tracking events have the `gla_mc_id` property and correct ID value after a GMC account is connected.
7. Proceed to the last step of onboarding flow.
8. Check if all tracking events don't have the `gla_ads_id` property before connecting a Google Ads account.
9. Check if all tracking events have the `gla_ads_id` property and correct ID value after a Google Ads account is connected.
10. Go to the Product Feed page.
11. Enable the "Preserve log" option in the Console tab.
12. Click one of the "Edit" links in the Product Feed table.
13. Wait for the product editing page to load.
14. Check if the queued tracking event `gla_edit_product_click` has the base properties.
   ![1](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/ee8719e8-1c62-4069-91d1-6e850675f47b)
15. Randomly operate the UI to trigger more tracking events to see if all have the base properties.
16. Global search the codebase to check if any `recordEvent` or `queueRecordEvent` call is missed to replace.

### Changelog entry

> Add - Include plugin version, Google Merchant Center account ID, and Google Ads account ID in all frontend tracking events.
